Sources/AsyncQueue/ActorQueue.swift

```diff
@@ -99,18 +78,90 @@ public final class ActorQueue: Sendable {
+        }
+    }
+
+    /// Schedules a synchronous task for execution and immediately returns.
+    /// The schedueled task will not execute until all prior tasks have completed or suspended.
+    /// - Parameter task: The task to enqueue.
+    public func async(_ task: @escaping @Sendable () -> Void) {
+        taskStreamContinuation.yield(.sync(task))
+    }
+
+    /// Schedules an throwing task and returns after the task is complete.
+    /// The schedueled task will not execute until all prior tasks have completed or suspended.
+    /// - Parameter task: The task to enqueue.
+    /// - Returns: The value returned from the enqueued task.
+    public func await<T>(_ task: @escaping @Sendable () -> T) async -> T {
+        await withUnsafeContinuation { continuation in
+            taskStreamContinuation.yield(.sync({
+                continuation.resume(returning: task())
+            }))
+        }
+    }
+
+    /// Schedules an task and returns after the task is complete.
+    /// The schedueled task will not execute until all prior tasks have completed or suspended.
+    /// - Parameter task: The task to enqueue.
+    /// - Returns: The value returned from the enqueued task.
+    public func await<T>(_ task: @escaping @Sendable () throws -> T) async throws -> T {
+        try await withUnsafeThrowingContinuation { continuation in
+            taskStreamContinuation.yield(.sync({
+                do {
+                    continuation.resume(returning: try task())
+                } catch {
+                    continuation.resume(throwing: error)
+                }
+            }))
         }
     }
 
     // MARK: Private
 
-    private let streamTask: Task<Void, Never>
-    private let taskStreamContinuation: AsyncStream<@Sendable () async -> Void>.Continuation
+    private let taskStreamContinuation: AsyncStream<TaskType>.Continuation
+
+    // MARK: - TaskType
+
+    enum TaskType {
+        case sync(@Sendable () -> Void)
+        case async(@Sendable () async -> Void)
+    }
+
+    // MARK: - ActorExecutor
+
+    private actor ActorExecutor {
+        func suspendUntilStarted(_ task: TaskType) async {
+            let semaphore = Semaphore()
+            executeWithoutWaiting(task, afterSignaling: semaphore)
+            // Suspend the calling code until our enqueued task starts.
+            await semaphore.wait()
+        }
+
+        private func executeWithoutWaiting(
+            _ task: TaskType,
+            afterSignaling semaphore: Semaphore)
+        {
+            // Utilize the serial (but not FIFO) Actor context to execute the task without requiring the calling method to wait for the task to complete.
+            Task {
+                switch task {
+                case let .sync(task):
+                    task()
+                    // Synchronous tasks can not re-enter this queue, so it is safe to wait until the task completes prior to signaling the semaphore.
+                    await semaphore.signal()
```

Comment:
`test_async_sendsSynchronousEventsInOrder` fails unless I `signal` the semaphore after the `task()` completes. I don't fully understand why just yet.

This task is executing within a serial actor context, so the `task` should complete within the same serial context as the semaphore signal. My best guess at an explanation is that signaling the semaphore requires suspension, and so ordering post-suspension to interact with the semaphore has no guaranteed order. However, I don't know how that hypothesis jives with the fact that the `.async`  case below works (and has worked on over 15mil repetitions).

I would love to better understand this behavior. Insights are appreciated!

As an aside: I am very much open to the feedback that this `.sync` case should exist on this queue at all. I added it to start a discussion around suspension, and am not convinced that this API should belong on the queue.

Reply:
> As an aside: I am very much open to the feedback that this `.sync` case should exist on this queue at all. I added it to start a discussion around suspension, and am not convinced that this API should belong on the queue.

Based on the intended usage as show in https://github.com/dfed/swift-async-queue/pull/3/files#r1033055297 it seems like there would never be a need for the synchronous methods, right?

Reply:
> `test_async_sendsSynchronousEventsInOrder` fails unless I `signal` the semaphore after the `task()` completes. I don't fully understand why just yet.
> 
> This task is executing within a serial actor context, so the `task` should complete within the same serial context as the semaphore signal. My best guess at an explanation is that signaling the semaphore requires suspension, and so ordering post-suspension to interact with the semaphore has no guaranteed order. However, I don't know how that hypothesis jives with the fact that the `.async`  case below works (and has worked on over 15mil repetitions).
> 
> I would love to better understand this behavior. Insights are appreciated!

I feel like the behavior is due to this code. https://github.com/dfed/swift-async-queue/blob/17894cc8e356de009fe93d3855715d47ab0b4794/Sources/AsyncQueue/ActorQueue.swift#L42-L47

If we signal the semaphore before executing the task then that code can begin processing the next task before the prior task has been completed. And the next task can be processed in parallel and is thus racing the previous task.

Reply:
I am also a bit confused initially why the `.async` case passes when we signal before executing the task.

Reply:
I think the `.async` case works because the task closure in that case is async itself, so it gets ordering from the actor. In the `.sync` case the task closure is not async so there is nothing enforcing the order of execution. I agree this is pretty trippy to think through 😵

Reply:
> it seems like there would never be a need for the synchronous methods, right?

There's not a need, no. But it could be a nice-to-have, if confusing feature. I'll remove the `.sync` option on my next pass.

Reply:
> If we signal the semaphore before executing the task then that code can begin processing the next task before the prior task has been completed. And the next task can be processed in parallel and is thus racing the previous task.

Yes, but the `executor` in the code you linked is an actor, and we're already processing code within the actor so it shouldn't be possible for the next task to begin execution on the actor context...

And of course as I typed that out I realized it's not possible for the next task to begin execution on the actor context _until the actor context suspends_, and we are suspending when we interact with the semaphore.

Reply:
> I think the `.async` case works because the task closure in that case is async itself, so it gets ordering from the actor. In the `.sync` case the task closure is not async so there is nothing enforcing the order of execution. I agree this is pretty trippy to think through 😵

This is my understanding as well, though I wish I better understood what was happening under the hood. It feels like the code I've written works by accident rather than intention. But I've run the async ordering test case 20+million times and it has never failed...

Reply:
It's certainly going to take me a little while to establish a mental model for Swift concurrency.

Reply:
With 10ec535 I think I have a much better understanding of how this works.

We are blocking the calling async for-loop until we've entered a `Task` running within an `actor`'s context. When we suspend to interact with another `actor` within the `await task()` line, we enqueue work for that other `actor` in order.

If the `task()` closure is synchronous, the `await` triggers suspension but since we don't have ordering from interacting with another actor, the suspension leads to random ordering.

The reason we were seeing `test_async_sendsSynchronousEventsInOrder` fail unless I `signal`d the semaphore until after the `task()` completed (see the OP on this thread) was because of [this bug](https://github.com/apple/swift/issues/62503) – adding the `_ = void` trick seen in 10ec535 enabled me to get `test_async_sendsSynchronousEventsInOrder` passing when `signal`ing the semaphore _before_ executing the `task()`.

Once this PR lands I think I'll try to enable adding synchronous tasks back to the `ActorQueue` (though I'm still waffling on it). It could be useful to ensure ordering of events enqueued from the isolated context and nonisolated contexts. We’ll Get There™, maybe.

Since we've now got a good handle on what was happening here, I'm resolving this thread.